### DISCUSSION TITLE
Short & Long term Roadmap

### DISCUSSION THREAD

**Discussion by sedefsavas** [kubernetes-sigs/cluster-api-provider-aws#2947] [Ideas]
What features do we want to add in the short (v1) and long term (v2+) to CAPA?

Please create an issue for each feature you propose and then add a brief summary below with the issue link. We can vote for them below to prioritize.

**Note 1**: Short term feature ideas should not have any breaking API changes.

**Note 2**: It will be great if you can add a note to the issues if you can contribute to its implementation. This will help us determine a more realistic roadmap.

**Comment by sedefsavas** [comment]
Parallelize resource creation for faster reconciliation: https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2953

**Comment by richardcase** [comment]
For EKS:

- bootstrap.sh options #2853
- pre-bootstrap commands #2771 
- Launch Templates #2055, #2224
- IPv6 and dual-stack support
- Stage 2 Identity Types #2336 
- EKS e2e improvements (multiple issues)
- ClusterClass for AWSManagedControlPlane (no issue yet)

**Comment by sedefsavas** [comment]
Supporting different network topologies: https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1484

**Comment by sedefsavas** [comment]
Spot instance support in MachinePools:https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2523

**Reply by sedefsavas** [reply]
Also, mix usage of On-Demand and Spot Instances in ASG is supported.

**Comment by sedefsavas** [comment]
AWS VPC CNI support for Cluster API managed clusters:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/3286

**Comment by sedefsavas** [comment]
IPv6 only and dual stack support: https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2420
https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/3381

**Comment by sedefsavas** [comment]
Add workload identity support to CAPA-managed clusters, similar to EKS IRSA (IAM roles for service accounts) support (EKS-managed CAPA clusters have this support)
https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2054

**Reply by sedefsavas** [reply]
A rough EKS IRSA workflow:

- EKS creates OIDC provider
- Using [service account token volume projection](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection), JWT token is requested from the OIDC provider by kubelet and the token is automatically updated.
- This token is mounted to the pods by kubelet
- Use webhooks to set AWS_WEB_IDENTITY_TOKEN_FILE and AWS_ROLE_ARN environment variables in the pods.

**Comment by sedefsavas** [comment]
Use AWS Controllers for Kubernetes (ACK) for AWS SDK calls:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/discussions/3306

**Comment by sedefsavas** [comment]
Multiple VPC CIDR support/ Secondary CIDR support
https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/3287

**Comment by sedefsavas** [comment]
Graceful shutdown of nodes: https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2574

2 cases to cover under this feature:
1. Draining MachinePool nodes during ASG scale-down.
2. Draining spot instances in MachinePools and MDs (using 2 minutes of warning time to drain the nodes)

AWS node termination handler has support for both cases above. So we can design these together.

**Reply by sedefsavas** [reply]
Related cluster-api proposal for spot instance termination handling:
https://github.com/kubernetes-sigs/cluster-api/pull/3528

TL;DR
Interruptible label is added to the spot instance nodes by Machine controller by checking AWSMachine.Status.Interruptible field: https://github.com/kubernetes-sigs/cluster-api/pull/3668

Using the interruptible labels on the nodes, we can install termination handler as DaemonSet:
https://github.com/aws/aws-node-termination-handler

**Reply by sedefsavas** [reply]
Also, once[ MachinePool Machines proposal](https://github.com/kubernetes-sigs/cluster-api/pull/6088) is in, when shut-down signal of a node is detected, we can utilize MHC to clean up resources following this proposal: https://github.com/kubernetes-sigs/cluster-api/pull/3528

**Comment by sedefsavas** [comment]
Validate **Amazon FSx for Lustre storage CSI** and **EFS CSI** plugins, similar to existing EBS CSI validation.

**Comment by sedefsavas** [comment]
Support for **AWS EC2 Fleet**, which
-  allows grouping mixed nodes by specifying different instance types with on-demand and spot purchase models
- launches the lowest price combination of instances to meet the target capacity based on these preferences

**Comment by sedefsavas** [comment]
Support for **AWS Spot Fleet**, which by selecting multiple instance types and max price, creates spot instances to meet target capacity.

**Comment by richardcase** [comment]
@sedefsavas - a great list :+1:

**Comment by pydctw** [comment]
Support cluster autoscaler

**Reply by sedefsavas** [reply]
AFAIK cluster autoscaler works as is with MachineDeployments.
For MachinePool support, MachinePool Machine proposal must be implemented.
If there is more to it other than adding an e2e test, would be good to have an issue.

**Reply by pydctw** [reply]
Agreed. I looked into the cluster autoscaler in my previous project and adding annotations on MD will do. This item is more for e2e test.

**Reply by richardcase** [reply]
And karpenter

**Comment by sedefsavas** [comment]
Windows nodes support
